File: J9Ch/src/S_1_Follow_up_in_CodeInterview/Required_5/_406_Minimum_Size_Subarray_Sum_Medium.java
package S_1_Follow_up_in_CodeInterview.Required_5;
import java.util.*;
/** 406 Minimum Size Subarray Sum


 * Created by tianhuizhu on 6/28/17.
 */
public class _406_Minimum_Size_Subarray_Sum_Medium {

    public class Solution {
        /**
         * @param nums: an array of integers
         * @param s: an integer
         * @return: an integer representing the minimum size of subarray
         */
        public int minimumSize(int[] nums, int s) {
            // write your code here
            int j = 0, i = 0;
            int sum =0;
            int ans = Integer.MAX_VALUE;
            for(i = 0; i < nums.length; i++) {
                while(j < nums.length && sum < s) {
                    sum += nums[j];
                    j ++;
                }
                if(sum >=s) {
                    ans = Math.min(ans, j - i);
                }
                sum -= nums[i];
            }
            if(ans == Integer.MAX_VALUE)
                ans = -1;
            return ans;
        }
    }
}

File: J9Ch/src/S_7_Follow_Up/Required_14/_406_Minimum_Size_Subarray_Sum_Medium.cpp

class Solution {
public:
    /**
     * @param nums: a vector of integers
     * @param s: an integer
     * @return: an integer representing the minimum size of subarray
     */
    int minimumSize(vector<int> &nums, int s) {
        // write your code here
        int n = nums.size();
        if (n==0) return -1;
        int left = 0, right = 0, total = 0, ans = n+1;
        while (right<n) {
            while (right<n && total<s) total += nums[right++];
            if (total<s) break;
            while (left<right && total>=s) total -= nums[left++];
            ans = min(ans, right-left+1);
        }
        if (ans==n+1) return -1;
        else return ans;
    }
};